docs/adr/0007-azure-pipeline-security-model.md

```diff
@@ -0,0 +1,70 @@
+# Azure pipeline security model
+
+## Status
+
+Proposed
+
+
+## Context
+
+This ADR defines an approach for a least-privilege Azure DevOps Pipeline that supports the automated deployment, teardown and re-deployment of a complete Marain instance.
+
+The challenge here is to balance the varied benefits of having a self-sufficient, repeatable pipeline with the need for adequate security controls to minimise the risk of the pipeline becoming the source of a security incident.
+
+Such a deployment pipeline needs to be able to perform the following actions:
+
+1. Create resources in the target subscription
+1. Query AAD for existing applications & service principals
+1. Perform ARM role assignments in the target subscription
+1. Create and manage new AAD applications & service principals
+
+### Create resources in the target subscription
+By default, the 'Service Connections' that Azure Pipelines use to authenticate with Azure only provide #1 of the above.
+
+### Perform ARM role assignments in the target subscription
+Typically this is achieved by granting the identity behind the service connection the 'owner' role in the target subscription.
+
+### Query AAD for existing applications & service principals
+The main pipeline is restructured to expect the required information to be provided as manually-maintained configuration.
+
+### Create and manage new AAD applications & service principals
+This is commonly handled in one of three ways:
```

Comment:
Worth noting that this is generally a one-time operation. That's significant because it means that moving it out of the pipeline isn't necessarily a big problem. (Particularly if you have one-time jobs that need doing when you first stand up an environment. E.g., if I'm remembering correctly that you can grant the ADO SP ownership only of the resource groups and not the whole sub, then you now have a one-time job of creating those resource groups and assigning ownership. If you're already on the hook for needing such a one-time procedure that needs higher privileges than the ADO SP will need in the long run, you may was well move other higher-privilege steps into that as well.)

Reply:
In my experience 'one-time' operations end up being either sufficiently not one-time (e.g you want a complete tear-down for a full 'proof of process' test) or sufficiently regular that the friction from it not being part of an automated process leads to undesirable 'workarounds' by those having to put up with it.

A self-service automated process for those higher-privilege steps, might of course be the ideal solution but that doesn't exist today - at least not off-the-shelf.

I should also add something about the self-healing benefits that the automated approach offers.